modules/oe_subscriptions_anonymous/oe_subscriptions_anonymous.tokens.inc

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/**
+ * @file
+ * Token callbacks for the Anonymous subscriptions module.
+ */
+
+declare(strict_types=1);
+
+use Drupal\Core\Render\BubbleableMetadata;
+use Drupal\Core\Url;
+
+/**
+ * Implements hook_token_info().
+ */
+function oe_subscriptions_anonymous_token_info(): array {
+  $type = [
+    'name' => t('Anonymous subscriptions'),
+    'description' => t('Tokens related to anonymous subscriptions.'),
+  ];
+
+  $anonymous_subscriptions['subscriptions_page'] = [
+    'name' => t('Subscriptions page'),
+    'description' => t('Link to subscriptions page based on message owner, to be used in message entities.'),
+    'needs-data' => 'message',
+  ];
+
+  return [
+    'types' => ['oe_subscriptions_anonymous' => $type],
+    'tokens' => ['oe_subscriptions_anonymous' => $anonymous_subscriptions],
+  ];
+}
+
+/**
+ * Implements hook_tokens().
+ */
+function oe_subscriptions_anonymous_tokens(string $type, array $tokens, array $data, array $options, BubbleableMetadata $bubbleable_metadata) {
+  $replacements = [];
+
+  if (
+    $type == 'oe_subscriptions_anonymous' &&
+    isset($tokens['subscriptions_page']) &&
+    !empty($data['message'])
+  ) {
+    /** @var \Drupal\message\Entity\Message $message */
+    $message = $data['message'];
+    $route = 'oe_subscriptions_anonymous.user_subscriptions.request_access';
+    $options = [
+      'absolute' => TRUE,
+    ];
+
+    if ($message->getOwner() !== NULL && $message->getOwner()->isCoupled()) {
```

Comment:
We are adding a polymorphic token that works for both type of users. But if a developer enables only the main oe_subscriptions module, no token is added. The main module should provide the token that gives access to the `oe_subscriptions.user.subscriptions` route, and the anonymous module should change said token for decoupled users.

Reply:
Moved to main module, and anonymous alters the token